lib/pavilion/scheduler_plugins.py

```diff
@@ -1,39 +1,71 @@
 import collections
 from pavilion.variables import VariableSetManager
-from yapsy.IPlugin import IPlugin
+from yapsy import IPlugin
 import logging
 import re
 
 LOGGER = logging.getLogger('pav.{}'.format(__name__))
 
-class PluginSchedulerError(RuntimeError):
+class SchedulerPluginError(RuntimeError):
     pass
 
-_SCHEDULER_PLUGINS = {}
+_SCHEDULER_PLUGINS = None
+_LOADED_PLUGINS = None
+
+_SCHEDULER_VARS = [ 'num_nodes', 'procs_per_node', 'qos', 'reservation',
+                    'partition', 'account', 'down_nodes', 'unused_nodes',
+                    'busy_nodes', 'maint_nodes', 'other_nodes', 'chunk_size' ]
 
 class SchedVarDict( collections.UserDict ):
 
     def __init__( self ):
         global _SCHEDULER_PLUGINS
-        super().__init__( _SCHEDULER_PLUGINS )
+        if _SCHEDULER_PLUGINS is not None:
+            raise SchedulerPluginError(
+                  "Dictionary of scheduler plugins can't be generated twice." )
+        super().__init__( {} )
+        _SCHEDULER_PLUGINS = self
+
+    def __getitem__( self, name, var ):
+        if name not in self.data:
+            self.data[ name ][ var ] = \
+                             get_scheduler_plugin( name ).get( var )
+        return self.data[ name ]
+
+    def _reset( self ):
+        LOGGER.warning( "Resetting the plugins.  This functionality exists " +\
+                        "only for use by unittests." )
+        _reset_plugins()
+        self.data = {}
+
+def _reset_plugins():
+    global _SCHEDULER_PLUGINS
 
-    def __getitem__( self, name ):
-        plugin = self.data[ name ]
-        return plugin.get()
+    if _SCHEDULER_PLUGINS is not None:
+        for key in list(_SCHEDULER_PLUGINS.keys()):
+            remove_system_plugins( key )
 
 def add_scheduler_plugin( scheduler_plugin ):
     global _SCHEDULER_PLUGINS
     name = scheduler_plugin.name
 
-    if name not in _SCHEDULER_PLUGINS:
-        _SCHEDULER_PLUGINS[ name ] = scheduler_plugin
-    elif priority > _SCHEDULER_PLUGINS[name].priority:
-        _SCHEDULER_PLUGINS[ name ] = scheduler_plugin
-    elif priority == _SCHEDULER_PLUGINS[name].priority:
-        raise PluginSchedulerError("Two plugins for the same system plugin "
+    if _LOADED_PLUGINS is None:
```

Comment:
I'm not sure we need priority for schedulers, either. Priority is there for plugins the user is likely to muck with. Again, I don't think it hurts.

Reply:
I'm also not sure why your loaded plugins need to be separate from you magic dictionary of plugins. The magic dictionary was necessary for system variables because of how they were deferred. 

Oh, and I just realized the scheduler plugins need to be distinct from the scheduler classes. Each test needs a unique instance of the scheduler class, whereas activated plugins are the single instance of that class. You may have done this, I've yet to see.